File: native/android/VueNativeCore/src/main/kotlin/com/vuenative/core/Components/Factories/VPressableFactory.kt
package com.vuenative.core

import android.content.Context
import android.view.MotionEvent
import android.view.View
import com.google.android.flexbox.FlexboxLayout

/**
 * Factory for VPressable — a generic pressable container component.
 *
 * Like VButton but without built-in text/label support. Uses PressableView,
 * a subclass of TouchableView that adds pressIn/pressOut callbacks.
 */
class VPressableFactory : NativeComponentFactory {

    override fun createView(context: Context): View {
        return PressableView(context)
    }

    override fun updateProp(view: View, key: String, value: Any?) {
        val pressable = view as? PressableView
        if (pressable == null) {
            StyleEngine.apply(key, value, view)
            return
        }

        when (key) {
            "disabled" -> {
                pressable.isDisabled = value == true || value == "true" ||
                        (value is Number && value.toInt() != 0)
            }
            "activeOpacity" -> {
                pressable.activeOpacity = when (value) {
                    is Double -> value.toFloat()
                    is Float -> value
                    is Int -> value.toFloat()
                    is String -> value.toFloatOrNull() ?: 0.7f
                    else -> 0.7f
                }
            }
            else -> StyleEngine.apply(key, value, view)
        }
    }

    override fun addEventListener(view: View, event: String, handler: (Any?) -> Unit) {
        val pressable = view as? PressableView ?: return

        when (event) {
            "press" -> pressable.onPress = { handler(null) }
            "longPress", "longpress" -> pressable.onLongPress = { handler(null) }
            "pressIn", "pressin" -> pressable.onPressIn = { handler(null) }
            "pressOut", "pressout" -> pressable.onPressOut = { handler(null) }
        }
    }

    override fun removeEventListener(view: View, event: String) {
        val pressable = view as? PressableView ?: return

        when (event) {
            "press" -> pressable.onPress = null
            "longPress", "longpress" -> pressable.onLongPress = null
            "pressIn", "pressin" -> pressable.onPressIn = null
            "pressOut", "pressout" -> pressable.onPressOut = null
        }
    }

    override fun insertChild(parent: View, child: View, index: Int) {
        val flex = parent as? FlexboxLayout ?: return
        val lp = StyleEngine.buildFlexLayoutParams(child)
        if (index >= flex.childCount) {
            flex.addView(child, lp)
        } else {
            flex.addView(child, index, lp)
        }
    }

    override fun removeChild(parent: View, child: View) {
        (parent as? FlexboxLayout)?.removeView(child)
    }
}

/**
 * Extension of TouchableView that adds pressIn and pressOut callbacks.
 * Fires pressIn when a touch begins and pressOut when the touch ends or is cancelled.
 */
class PressableView(context: android.content.Context) : TouchableView(context) {

    /** Called when a touch begins inside the view bounds. */
    var onPressIn: (() -> Unit)? = null

    /** Called when a touch ends or is cancelled. */
    var onPressOut: (() -> Unit)? = null

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (isDisabled) return false

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                onPressIn?.invoke()
            }
            MotionEvent.ACTION_UP -> {
                onPressOut?.invoke()
            }
            MotionEvent.ACTION_CANCEL -> {
                onPressOut?.invoke()
            }
        }

        return super.onTouchEvent(event)
    }
}

File: native/macos/VueNativeMacOS/Sources/VueNativeMacOS/Components/Factories/VPressableFactory.swift
import AppKit
import ObjectiveC

/// Factory for VPressable — custom mouse-tracking view for pressable interactions.
/// Maps to a PressableView (ClickableView subclass) with press, longPress,
/// pressIn (mouseDown), and pressOut (mouseUp) events.
final class VPressableFactory: NativeComponentFactory {

    // MARK: - Associated object keys

    private static var pressInHandlerKey: UInt8 = 0
    private static var pressOutHandlerKey: UInt8 = 0

    // MARK: - NativeComponentFactory

    func createView() -> NSView {
        let pressable = PressableView()
        pressable.ensureLayoutNode()
        return pressable
    }

    func updateProp(view: NSView, key: String, value: Any?) {
        guard let pressable = view as? PressableView else {
            StyleEngine.apply(key: key, value: value, to: view)
            return
        }

        switch key {
        case "disabled":
            if let disabled = value as? Bool {
                pressable.isDisabled = disabled
            } else if let disabled = value as? Int {
                pressable.isDisabled = disabled != 0
            } else {
                pressable.isDisabled = false
            }

        case "activeOpacity":
            if let opacity = value as? Double {
                pressable.activeOpacity = CGFloat(opacity)
            } else if let opacity = value as? Int {
                pressable.activeOpacity = CGFloat(opacity)
            } else {
                pressable.activeOpacity = 0.7
            }

        default:
            StyleEngine.apply(key: key, value: value, to: view)
        }
    }

    func addEventListener(view: NSView, event: String, handler: @escaping (Any?) -> Void) {
        guard let pressable = view as? PressableView else { return }

        switch event {
        case "press":
            pressable.onPress = {
                handler(nil)
            }

        case "longPress":
            pressable.onLongPress = {
                handler(nil)
            }

        case "pressIn":
            pressable.onPressIn = {
                handler(nil)
            }

        case "pressOut":
            pressable.onPressOut = {
                handler(nil)
            }

        default:
            break
        }
    }

    func removeEventListener(view: NSView, event: String) {
        guard let pressable = view as? PressableView else { return }

        switch event {
        case "press":
            pressable.onPress = nil
        case "longPress":
            pressable.onLongPress = nil
        case "pressIn":
            pressable.onPressIn = nil
        case "pressOut":
            pressable.onPressOut = nil
        default:
            break
        }
    }
}

// MARK: - PressableView

/// ClickableView subclass that adds pressIn (mouseDown) and pressOut (mouseUp) callbacks.
/// Provides the full set of press interaction events for the VPressable component.
final class PressableView: ClickableView {

    /// Called immediately when the mouse button goes down.
    var onPressIn: (() -> Void)?

    /// Called when the mouse button is released (regardless of position).
    var onPressOut: (() -> Void)?

    override func mouseDown(with event: NSEvent) {
        onPressIn?()
        super.mouseDown(with: event)
    }

    override func mouseUp(with event: NSEvent) {
        onPressOut?()
        super.mouseUp(with: event)
    }
}
